src/samples/audio/smart_amp_test_ipc4.c

```diff
@@ -447,178 +288,163 @@ static smart_amp_proc get_smart_amp_process(struct comp_dev *dev,
 	}
 }
 
-static int smart_amp_copy(struct comp_dev *dev)
+/* Only reset the src_input and fb_input if the num_input_buffers
+ * (get from module adapter)has changed.
+ */
+static void smart_amp_input_setup(struct processing_module *mod,
+				  struct input_stream_buffer *inputs,
+				  int32_t num_input_buffers)
 {
-	struct smart_amp_data *sad = comp_get_drvdata(dev);
-	struct comp_buffer __sparse_cache *source_buf = buffer_acquire(sad->source_buf);
-	struct comp_buffer __sparse_cache *sink_buf = buffer_acquire(sad->sink_buf);
-	uint32_t avail_passthrough_frames;
-	uint32_t avail_feedback_frames;
-	uint32_t avail_frames = 0;
-	uint32_t source_bytes;
-	uint32_t sink_bytes;
-	uint32_t feedback_bytes;
+	struct smart_amp_data *sad = module_get_private_data(mod);
+	struct comp_buffer __sparse_cache *buf;
+	int i;
 
-	comp_dbg(dev, "smart_amp_copy()");
+	if (sad->input_stream_num == num_input_buffers)
+		return;
 
-	/* available bytes and samples calculation */
-	avail_passthrough_frames =
-		audio_stream_avail_frames(&source_buf->stream,
-					  &sink_buf->stream);
+	sad->input_stream_num = num_input_buffers;
 
-	k_mutex_lock(&sad->lock, K_FOREVER);
-	if (sad->feedback_buf) {
-		struct comp_buffer __sparse_cache *buf = buffer_acquire(sad->feedback_buf);
+	if (num_input_buffers == 1) {
+		sad->src_input = &inputs[0];
+		sad->fb_input = NULL;
+		return;
+	}
 
-		if (buf->source && comp_get_state(dev, buf->source) == dev->state) {
-			/* feedback */
-			avail_feedback_frames =
-				audio_stream_get_avail_frames(&buf->stream);
+	for (i = 0; i < num_input_buffers; i++) {
+		buf = attr_container_of(inputs[i].data,
+					struct comp_buffer __sparse_cache,
+					stream, __sparse_cache);
 
-			avail_frames = MIN(avail_passthrough_frames,
-					   avail_feedback_frames);
+		if (IPC4_SINK_QUEUE_ID(buf->id) == SOF_SMART_AMP_FEEDBACK_QUEUE_ID)
+			sad->fb_input = &inputs[i];
+		else
+			sad->src_input = &inputs[i];
```

Comment:
We still disagree on this: I think it's better not to store pointers to acquired buffers in persistent objects, instead it isn't too large of an overhead to run the identification on each copy. @andrula-song thinks it's too much of an overhead and since we do own buffers where we use them, it's ok to also store them persistently. I agree, that in practice there's no difference - as long as we don't "accidentally" access those pointers outside of the `.process()` method, the damage is "only theoretical." Also this is "just a sample," not the real production code. So, if others agree, that this is ok, we can keep it this way. At least I'd put a comment here.

Reply:
@lyakh while this is sample code, the real smart amp implementation will have to follow this lead as well. I am also worried that the production code might even involve multiple cores. Not sure if it will make the situation worse or not but I tihnk it's best to avoid it. I have an alternate suggestion. We do have the bind/unbind ops in the module adapter where we save the sources in a list. In a way this list in ordered in terms of the sources getting added as the modules are bound. So in the case of the smart amp, the source component for the playback path will always be at the 0th index and the feedback comp if available will always be at 1. Can we make use of this info to retrieve the source buffer from the 0th index and feedback buffer from the first index during process? And if w used the module_source_info_acquire/module_source_info_release() in the smart amp module, we'd also avoid coherency issues. what do you think?

Reply:
@ranj063 I would be cautious using the module binding order. Why is it fixed? Can it change? We were thinking about a softer approach - during initialisation (binding or whatever step is suitable) check which index corresponds to the source (or to the feedback) buffer and record the index. But even about that I cannot say how reliable it would be. Can one of them (e.g. index 0) be unbound at run-time and then re-bound this time as index 0?

Reply:
@ranj063 ok, so you're saying, that that's already used there. Ok, then we won't break it further if we use that information again. But let's make it explicit somewhere - put a check that that's indeed the case somewhere to be run where buffers can potentially be reconfigured and a comment, can that be done?

Reply:
@ranj063  @lyakh 
@lyakh if the order is not fixed why is it that when num_input_buffers == 0, that it is always the src buffer and not the feedback buffer?
 >> in my opinion, if num_input_buffers == 0, that it is always the src  because the smart_amp is in playback pipeline. if there is only feedback (capture pipeline), we even never create smart_amp this component.

Reply:
@ranj063 Actually the last push of my code already used module_source_info_acquire/release to protect the bind/unbind interrupt during smart_amp processing(though I met a exception, but I agree to use the acquire/release to keep it safer). But what's your thought about getting the source and feedback buffers from the module source info array? Is it to use the ipc_config.type? Why don't we use the buffer id since we have this info in IPC4. It seems very tricky to determinate the input by info outside of smart_amp itself.

Reply:
@lyakh i just had a chat with @andrula-song  and what she's storing in sad src_input and fb_input is not a pointer to the acquired buffer or the streams but rather the pointer to the input_buffers in struct processing_module. So this part is actually fine.

Reply:
> Actually the last push of my code already used module_source_info_acquire/release to protect the bind/unbind interrupt during smart_amp processing

just a note: `module_source_info_acquire()` doesn't actually protect anything unless multiple cores are used for some of the modules inputs or outputs, i.e. in most cases it simply converts an uncached address to a cached one. Also note, that objects, used with the coherent API should use that API consistently for all accesses to the data. Mixing cached and uncached access to the data (i.e. with and without `module_source_info_acquire()`) will cause data corruption

Reply:
> @lyakh i just had a chat with @andrula-song and what she's storing in sad src_input and fb_input is not a pointer to the acquired buffer or the streams but rather the pointer to the input_buffers in struct processing_module. So this part is actually fine.

@ranj063 yes, currently those `input_buffers[]` are persistent but once we fix and get something like #7504 in, those will be on-stack objects, so storing them persistently will be a bug.

Reply:
ok, let's look up every processing. Now I know why you ask for such change, thanks.